onnxruntime/core/framework/execution_frame.cc

```diff
@@ -86,13 +86,54 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
                     " Requested shape:", shape ? shape->ToString() : "null");
       }
     } else {
+      VerifyOutputSizes(output_index, node, shape);
       status = CreateNodeOutputMLValueImpl(*p_ort_value, ort_value_idx, shape, nnz);
     }
   }
 
   return status;
 }
 
+void IExecutionFrame::VerifyOutputSizes(int output_index, const onnxruntime::Node& node,
+                                       const onnxruntime::TensorShape* output_shape) {
+  onnxruntime::ProtoHelperNodeContext protoContext(node);
+  onnxruntime::OpNodeProtoHelper<onnxruntime::ProtoHelperNodeContext> info(&protoContext);
+
+  const onnx::TypeProto* outputproto = info.GetOutputType(output_index);
+
+  if (outputproto == nullptr) {
+    return;
+  }
+
+  if (outputproto->value_case() != onnx::TypeProto::kTensorType) {
+    return;
+  }
+  const auto& tensortype = outputproto->tensor_type();
+  if (tensortype.has_shape()) {
+    const auto& shape = tensortype.shape();
+    if (shape.dim_size() != int(output_shape->NumDimensions())) {
+      if (GetLogger() != NULL) {
+        LOGS(*GetLogger(), WARNING) << "Number of dimension of output shape didn't match "
+                                               << "with model's expected number of dimension of output shape";
+      }
+      return;
+    }
+
+    for (uint32_t output_dim = 0; output_dim < output_shape->NumDimensions(); ++output_dim) {
+      if (shape.dim(output_dim).has_dim_value()) {
+        int64_t expected_size = shape.dim(output_dim).dim_value();
```

Comment:
Can avoid calling shape.dim(output_dim) the second time.

Reply:
Acknowledged.